portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,19 +14,101 @@
+import com.google.gson.Gson;
+import com.google.sps.data.Comment;
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+// import javax.servlet.ServletException;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that handles data of comment section. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
-  }
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        PreparedQuery results = datastore.prepare(query);
+
+        List<Comment> comments = new ArrayList<>();
+        for (Entity entity : results.asIterable()) {
+            long id = entity.getKey().getId();
+            String data = (String) entity.getProperty("data");
+            String userEmail = (String) entity.getProperty("userEmail");
+            long timestamp = (long) entity.getProperty("timestamp");
+
+            Comment comment = new Comment(id, data, userEmail, timestamp);
+            comments.add(comment);
+        }
+        
+        Gson gson = new Gson();
+
+        response.setContentType("application/json;");
+        response.getWriter().println(gson.toJson(comments));
+    }
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        UserService userService = UserServiceFactory.getUserService();
+
+        if (!userService.isUserLoggedIn()) {
+            response.sendRedirect("/index.html");
+            return;
+        }
+        
+        String data = request.getParameter("comment");
+        String userEmail = userService.getCurrentUser().getEmail();
+
+        if (checkValid(data)) {
+            long timestamp = System.currentTimeMillis();
+
+            Entity commentEntity = new Entity("Comment");
+            commentEntity.setProperty("data", data);
+            commentEntity.setProperty("userEmail", userEmail);
+            commentEntity.setProperty("timestamp", timestamp);
+
+            DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+            datastore.put(commentEntity);
+        }
+
+        response.sendRedirect("/index.html");
+    }
+
+    /** Checks if comment is valid or not. */
+    private boolean checkValid(String data) {
+        boolean valid = true;
```

Comment:
TLDR, this should be false.

fallback should always be safe. Assume invalid till you have ensured that the comment meets all validation conditions